[`pylint`] Implement `self-cls-assignment` (`W0642`)

diceroll123 opened the pull request:
## Summary

Implements [`W0642`/`self-cls-assignment`](https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/self-cls-assignment.html)

See: #970 

## Test Plan

`cargo test`

MichaReiser commented:
The rule fits into the suspicious category. We should consider renaming the rule to `self-or-cls-assignments` to make it grammatically correct `allow(self-or-cls-assignments)`

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+
+impl Violation for SelfClsAssignment {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let SelfClsAssignment { keyword } = self;
+        format!("Assignment of variable `{keyword}`")
+    }
```
I think we can improve the message here depending on whether `self` or `cls` was used. We can frame it in a similar way as Pylint does:

```
Invalid assignment to (self|cls) in (instance|class) method
```

We can use the `MethodType` enum as suggested below.

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for assignment of `self` and `cls` in methods.
+///
+/// ## Why is this bad?
+/// The identifiers `self` and `cls` are conventional in Python for
+/// the first argument of instance methods and class methods, respectively.
+///
```
Can you expand the documentation to include an "Example" section and optionally a "References" section? For example,

https://github.com/astral-sh/ruff/blob/de46a36bbcdc1df4a56fc486ae5c7bcf2c4bbba0/crates/ruff_linter/src/rules/flake8_2020/rules/name_or_attribute.rs#L20-L37

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+/// the first argument of instance methods and class methods, respectively.
+///
+#[violation]
+pub struct SelfClsAssignment {
```
As suggested by Micha, can you rename this to `self_or_cls_assignment`?

```suggestion
pub struct SelfOrClsAssignment {
```

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+}
+
+/// PLW0127
+pub(crate) fn self_cls_assignment(checker: &mut Checker, target: &Expr) {
```
As suggested by Micha, can you rename this to `self_or_cls_assignment`?
```suggestion
pub(crate) fn self_or_cls_assignment(checker: &mut Checker, target: &Expr) {
```

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+            let Some(first_arg) = parameters.args.first() else {
+                return;
+            };
+            if first_arg.parameter.name.as_str() != "cls" {
+                return;
+            }
```
We'd need to first check in `posonlyargs` and then in `args` as in:

```rs
    let Some(ParameterWithDefault {
        parameter: self_or_cls,
        ..
    }) = parameters
        .posonlyargs
        .first()
        .or_else(|| parameters.args.first())
    else {
        return;
    };
```

This is in case there's a `/` in the parameter list `def foo(self, foo, /): ...`.

dhruvmanila replied:
I went ahead and added test cases for this scenario

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+            "cls"
+        }
+        function_type::FunctionType::Method { .. } => {
+            let Some(first_arg) = parameters.args.first() else {
+                return;
+            };
+            if first_arg.parameter.name.as_str() != "self" {
+                return;
+            }
```
nit: I'd move these checks before the function classification above and possibly use an enum. We can then use this enum in `SelfOrClsAssignment` and accordingly format the error message as suggested earlier:

```rs
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum MethodType {
	Instance,
	Class,
}

impl MethodType {
	fn arg_name(self) -> &'static str {
		match self {
			MethodType::Instance => "self",
			MethodType::Class => "cls",
		}
	}
}
```

dhruvmanila commented on crates/ruff_linter/src/rules/pylint/rules/self_cls_assignment.rs:
```diff
@@ -0,0 +1,104 @@
+                }
+            }
+        }
+        Expr::Tuple(ast::ExprTuple { elts, .. }) => {
```
We'd also need to check in `Expr::List` in a similar manner as that's a valid assignment target.

dhruvmanila commented on crates/ruff_linter/resources/test/fixtures/pylint/self_cls_assignment.py:
```diff
@@ -0,0 +1,18 @@
+        self, blah = "red", "blue"  # PLW0642
+        blah, (self, blah2) = "apple", ("orange", "banana")  # PLW0642
+    
+    def ok(self) -> None:
+        cls = None  # OK because the rule looks for the name in the signature
```
Can you add a similar test case for `self` used in a class method?

dhruvmanila commented on crates/ruff_linter/resources/test/fixtures/pylint/self_cls_assignment.py:
```diff
@@ -0,0 +1,18 @@
+        cls: Fruit = "apple"  # PLW0642
+        cls += "orange"  # PLW0642
+        cls, blah = "apple", "orange"  # PLW0642
+        blah, (cls, blah2) = "apple", ("orange", "banana")  # PLW0642
```
We should also add test cases when the assignment target is a list. Refer to my comment below.

dhruvmanila commented on crates/ruff_linter/resources/test/fixtures/pylint/self_cls_assignment.py:
```diff
@@ -0,0 +1,18 @@
+class Fruit:
+    @classmethod
+    def list_fruits(cls) -> None:
```
Can we add some test cases which uses other type of functions? For example:

Using a normal function:
```py
def list_fruits(self, cls) -> None:
    self = "apple"  # Ok
    cls = "banana"  # Ok
```

Using a static method:
```py
class Fruit:
	@staticmethod
	def list_fruits(self, cls) -> None:
	    self = "apple"  # Ok
	    cls = "banana"  # Ok
```

dhruvmanila reviewed (changes requested):
Thank you for the PR!

This looks like a great start. I've a few recommendations around testing and logic but otherwise looks good.

diceroll123 commented:
Couldn't find any official Python documentation on why this is not a good thing to do in code, but I've added an example, and addressed other changes as well. 😄

dhruvmanila reviewed (approved):
Thank you! I made some additional changes which I've listed down:
* Check starred expressions (`*self = "foo"`)
* Update example case and add "Use instead:" section

Looking at the ecosystem changes, a follow-up would probably be to allow type casting like the following:

```py
self = cast(Fruits, self)
cls = cast(Colors, cls)
```
